src/Data/SemVer/Types.hs

```diff
@@ -195,22 +205,61 @@ renderSV = pack . show
 -- | Returns whether a given semantic version matches a range.
 -- Note that there are special cases when there are prerelease tags. For
 -- details see https://github.com/npm/node-semver#prerelease-tags.
--- matches :: SemVerRange -> SemVer -> Bool
--- matches range version = case (sharedTags range, svTags version) of
---   -- This is the simple case, where neither the range nor the version has given
---   -- prerelease tags. Then we can just do regular predicate calculus.
---   (Nothing, PrereleaseTags []) -> matchesSimple range version
---   _ -> undefined
---   -- (Just rTags, PrereleaseTags vTags)
---   --   | rTags == vTags -> matchesSimple range version
---   --   | tuplesOf range /= [toTuple version] -> False
---   --   | otherwise -> matchesTags range rTags vTags
---   -- (_, _) -> False
+matches :: SemVerRange -> SemVer -> Bool
+matches range version =
+  case (sharedTags range, svTags version) of
+
+    (Nothing, PrereleaseTags vTags)
+      -- Neither the range nor the version have prerelease tags
+      | null vTags -> matchesSimple range version
+
+      -- If there is no prerelease tag in the range but there is in
+      -- the version reject it
+      | otherwise  -> False
+
+    -- A range with a prerelease tag can match a version without a
+    -- prerelease tag provided it *does* meet the semantic version
+    -- tuple's constraint criteria
+    (Just _, PrereleaseTags []) ->
+      matchesSimple range version
+
+    -- The most important invariant when considering a comparison
+    -- between a range with prerelease tags and a version with
+    -- prerelease tags is whether the semantic version in both is the
+    -- same; if it is not, then we must reject the version.
+    --
+    -- Note that we could have a conjunction, so we want to see if our
```

Comment:
Might be helpful in this comment to show some examples. It's a little difficult to visualize the scenario